In [15]:
import numpy as np
import pandas as pd


# 1. Auto brochures

### pypdf

In [18]:
pip install PyPDF2

Note: you may need to restart the kernel to use updated packages.


In [19]:
import PyPDF2

In [20]:
file = open('Toyota_US-FullLine_2022.pdf', 'rb')

In [21]:
fileReader = PyPDF2.PdfFileReader(file)

In [22]:
print(fileReader.numPages)

33


In [23]:
fileReader.getPage(0)

{'/ArtBox': [0, 0, 792, 612],
 '/BleedBox': [0, 0, 792, 612],
 '/Contents': {'/Filter': '/FlateDecode'},
 '/CropBox': [0, 0, 792, 612],
 '/MediaBox': [0, 0, 792, 612],
 '/Parent': {'/Count': 5,
  '/Kids': [IndirectObject(499, 0),
   IndirectObject(1, 0),
   IndirectObject(13, 0),
   IndirectObject(28, 0),
   IndirectObject(60, 0)],
  '/Parent': {'/Count': 33,
   '/Kids': [IndirectObject(489, 0),
    IndirectObject(490, 0),
    IndirectObject(491, 0),
    IndirectObject(492, 0),
    IndirectObject(493, 0),
    IndirectObject(494, 0)],
   '/Type': '/Pages'},
  '/Type': '/Pages'},
 '/PieceInfo': {'/InDesign': {'/DocumentID': 'xmp.did:d65754be-b5a3-4301-9be4-e3809ae652f2',
   '/LastModified': 'D:20210813003003Z',
   '/NumberOfPageItemsInPage': 1,
   '/NumberofPages': 1,
   '/OriginalDocumentID': 'xmp.did:0fbc1d4e-ef8d-4b2e-8412-63c312b7d7c9',
   '/PageItemUIDToLocationDataMap': {'/0': [1656,
     0,
     3,
     -450,
     297,
     -342,
     333,
     1,
     0,
     0,
     1,
     -366

### PDFplumber

In [24]:
pip install  pdfplumber

Note: you may need to restart the kernel to use updated packages.


In [25]:
import pdfplumber
with pdfplumber.open(r'Toyota_US-FullLine_2022.pdf') as pdf:
    first_page = pdf.pages[1]
    print(first_page.extract_text())

02
Let the  
E
M
O
C
adventures  
L
E
W
begin.
CARS & MINIVAN  03
Behind the wheel of a Toyota, big and  
small adventures happen naturally.  
Whether you’re starting your career or   TRUCKS  18
introducing your newborn to the world,  
each day brings a new story. And wherever 
you’re going, smart safety technologies  
are designed to help protect you. Get ready  
to see where the road takes you next.  CROSSOVERS & SUVS  21
Cover images in clockwise order: Highlander XSE shown; RAV4 TRD Off-Road shown with available Technology Package; Venza Limited shown; Camry XSE Hybrid shown in Celestial Silver Metallic and Midnight Black 
TOYOTA.COM Metallic roof1 with available power tilt/slide moonroof. See numbered footnotes in Disclosures section.
Page 2


Package appears to work well :
https://github.com/jsvine/pdfplumber/tree/b904dd6a7c5164c5e051fa9b4121f8af6a105ae9

In [26]:
txt = pdfplumber.open(r'Toyota_US-FullLine_2022.pdf').pages[10].extract_text()

In [27]:
txt.split()

['Prius',
 'Prime',
 'Limited',
 'shown.',
 'PRIUS',
 'PRIME',
 'Charge',
 'up',
 'your',
 'Purposeful',
 'aerodynamics',
 'combined',
 'with',
 'an',
 'intelligent',
 'powertrain',
 'that',
 'produces',
 'versatile',
 'driving',
 'modes',
 'help',
 'you',
 'daily',
 'drive.',
 'get',
 'the',
 'most',
 'miles',
 'out',
 'of',
 'Prius',
 'Prime.',
 'TOYOTA.COM/PRIUSPRIME',
 'CARS',
 '&',
 'MINIVAN',
 '11',
 'Page',
 '11']

In [28]:
txt = pdfplumber.open(r'Volvo_US FullLine_2020.pdf').pages[2].extract_text()

In [29]:
txt

'T:12.25”\nVOLVO RANGE\nWHEN WE INNOVATE,\nTHE CAR IS NOT\nTOP OF MIND.\nWe admit it. You fascinate us. We think and talk about you all day. Everything  \nabout your life – its ups and downs, joys and disappointments, needs, demands,  \nfrustrations and hopes – captivates us. \nSo when we innovate, every detail – from service design right down to the  \nsmallest screw – has to justify itself to you. Does it add anything? Is it useful? \nWe know you always want to feel in control. Our IntelliSafe technologies not only   T\n:8\nhelp to save lives, but also give you more enjoyment behind the wheel.  ”\nWe know you want to stay connected to the world around you. Our Sensus  \nsolutions ensure you are. \nWe also know you care about the future of the planet. That’s why our Drive-E  \ntechnologies allow you to drive sustainably – without compromising on performance. \nAnd we know you’re interested in self-driving cars. Each day we take pioneering  \nsteps in our quest to deliver a safe and re

In [30]:
txt = pdfplumber.open(r'BMW_US Full_Line_2019.pdf').pages[2].extract_text()

In [31]:
txt.split()

['BMW',
 'X2',
 'BMW',
 'X3',
 'BMW',
 '7',
 'SERIES',
 'BMW',
 '530e',
 'PAST,',
 'PRESENT',
 'THE',
 'NEXT',
 '100',
 'YEARS',
 'BMW',
 'DESIGN',
 'ENGINEERING',
 'BMW',
 'INNOVATIONS',
 'AND',
 'FUTURE',
 'Future',
 'of',
 'BMW',
 'Hallmarks',
 'of',
 'performance',
 'Performance',
 'is',
 'paramount',
 'ConnectedDrive®',
 '|',
 'Gesture',
 'Control',
 '|',
 'Touch',
 'Display',
 'BMW',
 'M',
 'BMW',
 '2',
 'SERIES',
 'BMW',
 '3',
 'SERIES',
 'BMW',
 '4',
 'SERIES',
 'M240i',
 '|',
 'M2',
 '|',
 'M3',
 '|',
 'M4',
 '|',
 'M550i',
 'Coupe',
 '|',
 'Convertible',
 'Sedan',
 '|',
 'Sports',
 'Wagon',
 'Coupe',
 '|',
 'Convertible',
 'M5',
 '|',
 'M6',
 '|',
 'M760i',
 '|',
 'M850i',
 'Gran',
 'Turismo',
 'Gran',
 'Coupe',
 'X3',
 'M40i',
 '|',
 'X4',
 'M40i',
 '|',
 'X5',
 'M',
 '|',
 'X6',
 'M',
 'BMW',
 '5',
 'SERIES',
 'BMW',
 '6',
 'SERIES',
 'BMW',
 '7',
 'SERIES',
 'BMW',
 'i',
 'PERFORMANCE',
 'Sedan',
 'Gran',
 'Coupe',
 '|',
 'Gran',
 'Turismo',
 'Sedan',
 '330e',
 '|',
 '530e

### Conclusion
There is good signs with this project idea. But I see potential issues with the PDF text outputs. On the volvo brochure there are text formatting values like the following that may make it difficult to efficiently extract text on a larger scale:

H117005_P224010_Volvo_Range_MY20_Brochure_update_enUS_R7.indd\nSaved at 8-30-2019 12:22 PM from lcmmapwspp08524 by None / Gustavo Rodriguez Printed at None\nPage Count 3 of 20\nJob info Approvals Fonts & Images \nJob H117009_P224010 Art Director None Fonts: Volvo Novum (Light, Bold), Volvo Sans Pro (Light, Medium)\nClient Volvo Copywriter None\nImages: STE_090_41A4354_ret_prepress_CMYK.tif (CMYK; 551 ppi; 72.48%)\nMarket United States Account Mgr None\nBrand V40 Studio Artist Carina Cruz\nMedia Type None Production  Carla Ramírez\nLive 11.3852” x 7.1352” Proofreader Jon Mendoza\nTrim 12.25” x 8” Notes\nBleed None\nNone\nPubs None\nDocument Path LCMM_Print:Volvo:H117005:P224010:H117005_P224010_Volvo_Range_MY20_Brochure_update_enUS_R7.indd'

# 2 Youtube comments


In [3]:
 pip install --upgrade google-api-python-client

  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)
Note: you may need to restart the kernel to use updated packages.


In [58]:
# -*- coding: utf-8 -*-

# Sample Python code for youtube.commentThreads.list
# See instructions for running these code samples locally:
# https://developers.google.com/explorer-help/code-samples#python

import os

import googleapiclient.discovery


def main():
    # Disable OAuthlib's HTTPS verification when running locally.
    # *DO NOT* leave this option enabled in production.
    os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = "1"

    api_service_name = "youtube"
    api_version = "v3"
    DEVELOPER_KEY = "AIzaSyCBxycIEpc93U25rDCUONBoahy6WGFCmBU"

    youtube = googleapiclient.discovery.build(
        api_service_name, api_version, developerKey = DEVELOPER_KEY)

    request = youtube.commentThreads().list(
        part="id, snippet",
        maxResults=100,
        order="relevance",
        videoId="Mzj3_FjuDuI"
    )
    response = request.execute()

    return(response)



In [59]:
raw = main()

In [60]:
#### This format

'''
{
  "kind": "youtube#commentThreadListResponse",
  "etag": etag,
  "nextPageToken": string,
  "pageInfo": {
    "totalResults": integer,
    "resultsPerPage": integer
  },
  "items": [
    commentThread Resource
  ]
}

'''

'\n{\n  "kind": "youtube#commentThreadListResponse",\n  "etag": etag,\n  "nextPageToken": string,\n  "pageInfo": {\n    "totalResults": integer,\n    "resultsPerPage": integer\n  },\n  "items": [\n    commentThread Resource\n  ]\n}\n\n'

In [61]:
raw

{'kind': 'youtube#commentThreadListResponse',
 'etag': 'YPU-FmfZePieM8s5YNpuZ5gy7X4',
 'pageInfo': {'totalResults': 39, 'resultsPerPage': 100},
 'items': [{'kind': 'youtube#commentThread',
   'etag': 'vPhHA6gXLDZSJGj6LR4l0TROJUs',
   'id': 'UgxZxmn80aQsO-cVee54AaABAg',
   'snippet': {'videoId': 'Mzj3_FjuDuI',
    'topLevelComment': {'kind': 'youtube#comment',
     'etag': 'gizqCygsTI29zb3rJOcxc3a6iBM',
     'id': 'UgxZxmn80aQsO-cVee54AaABAg',
     'snippet': {'videoId': 'Mzj3_FjuDuI',
      'textDisplay': 'I just wanted to tell you that you are an excellent popularizer. You explain clearly and do it step by step. Thank you man!',
      'textOriginal': 'I just wanted to tell you that you are an excellent popularizer. You explain clearly and do it step by step. Thank you man!',
      'authorDisplayName': 'louisdavid',
      'authorProfileImageUrl': 'https://yt3.ggpht.com/ytc/AKedOLQ2uj09s3wRIU9HAK6vRsC_i0s7n5VwOSJE-wRN=s48-c-k-c0x00ffffff-no-rj',
      'authorChannelUrl': 'http://www.you

In [62]:
raw['items'][0]['snippet']['topLevelComment']['snippet']['textOriginal']

'I just wanted to tell you that you are an excellent popularizer. You explain clearly and do it step by step. Thank you man!'

In [63]:
raw['items'][2]['snippet']['topLevelComment']['snippet']['textOriginal']

'Great video and super clear explanation. Thank you for sharing.'

In [64]:
len(raw['items'])

39

### Conclusion

Really good start. Promising method that is straight foward to use.  The likely gathering procedure will be to choose videos.  Create a comment pull for each video and then create a dataframe from the comments within the nested dictionaries